File: shoring/lowerclass/src/main/java/com/panopset/lowerclass/VersionParser.java
package com.panopset.lowerclass;

import java.io.File;
import java.io.IOException;
import com.panopset.compat.Logop;
import com.panopset.compat.Stringop;
import com.panopset.compat.StatusListener;

public class VersionParser implements StatusListener {

  public static String DEFAULT_MAVEN_HOME = Stringop.USER_HOME + "/.m2";
  public static String PANOPSET_JAR = Stringop.USER_HOME + "/panopset.jar";

  public static void main(String... args) throws IOException {
    Logop.dspmsg("*** Entire repository example:");
    Logop.dspmsg(new VersionParser().getSummaryReport());
    Logop.dspmsg("*** Single jar example:");
    Logop.dspmsg(new VersionParser(PANOPSET_JAR).getDetailedReport());
  }

  public VersionParser() {}

  public VersionParser(final String pathToJar_or_directoryToTraverse) {
    this(new File(pathToJar_or_directoryToTraverse));
  }

  public VersionParser(final File jar_or_directoryToTraverse) {
    setFile(jar_or_directoryToTraverse);
  }

  private String mavenHome;

  private String getMavenHome() {
    if (mavenHome == null) {
      mavenHome = System.getenv().get("M2_HOME");
      if (mavenHome == null) {
        mavenHome = System.getenv().get("MAVEN_HOME");
      }
      if (mavenHome == null) {
        mavenHome = DEFAULT_MAVEN_HOME;
      }
    }
    return mavenHome;
  }

  private File file;

  private File getFile() {
    if (file == null) {
      file = new File(getMavenHome());
    }
    return file;
  }

  public void setFile(final File jarOrDirectory) {
    this.file = jarOrDirectory;
  }

  private String summaryReport;
  
  public String getSummaryReport() throws IOException {
    if (summaryReport == null) {
      summaryReport = createReport(false);
    }
    return summaryReport;
  }
  
  private String detailReport;

  public String getDetailedReport() throws IOException {
    if (detailReport == null) {
      detailReport = createReport(true);
    }
    return detailReport;
  }
 
  private String createReport(boolean printDetails) throws IOException {
    VersionMakeup vm = new VersionMakeup();
    vm.analyze(getFile(), printDetails);
    return vm.getReport();
  }

  public void update(String message) {
    Logop.dspmsg(message);
  }
}

File: code/shoring/desk/src/main/kotlin/com/panopset/desk/utilities/lowerclass/VersionParser.kt
package com.panopset.desk.utilities.lowerclass

import com.panopset.compat.Logz
import com.panopset.compat.StatusListener
import com.panopset.compat.Stringop
import java.io.File
import java.io.IOException

class VersionParser() : StatusListener {

    constructor(jarOrDirectoryToTraverse: File): this() {
        file = jarOrDirectoryToTraverse
    }

    private var mavenHome: String = ""
        get() {
            if (field.isEmpty()) {
                field = System.getenv()["M2_HOME"]?:""
                if (field.isEmpty()) {
                    field = System.getenv()["MAVEN_HOME"]?:""
                }
                if (field.isEmpty()) {
                    field = DEFAULT_MAVEN_HOME
                }
            }
            return field
        }

    private var file: File = File(mavenHome)

    var summaryReport: String = ""
        get() {
            if (field.isEmpty()) {
                field = createReport(false)
            }
            return field
        }
        private set

    private var detailReport: String? = null

    val detailedReport: String?
        get() {
            if (detailReport == null) {
                detailReport = createReport(true)
            }
            return detailReport
        }

    private fun createReport(printDetails: Boolean): String {
        val vm = VersionMakeup()
        vm.analyze(file, printDetails)
        return vm.report
    }

    companion object {
        var DEFAULT_MAVEN_HOME: String = "${Stringop.USH}/.m2"

        @Throws(IOException::class)
        @JvmStatic
        fun main(args: Array<String>) {
            Logz.dspmsg("*** Entire repository example:")
            Logz.dspmsg(VersionParser().summaryReport)
        }
    }

    override fun update(message: String) {
        Logz.dspmsg(message)
    }
}
